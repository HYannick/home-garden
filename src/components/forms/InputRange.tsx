/** @jsx jsx */
import React, {Fragment, useState} from 'react';
import styled from "@emotion/styled";
import {jsx} from '@emotion/core';
import {Range, getTrackBackground} from "react-range";

interface RangeProps {
  field?: any,
  form?: any,
  label?: string,
  min: number,
  max: number,
  step: number,
  onChange: Function
}

interface RailProps {
  isDragged: boolean
}

const FrequencyLabel = styled('label')`
  font-size: 1.6rem;
  color: ${({theme}) => theme.palette.grey.dark};
`;

const Rail = styled('div')`
  height: 3.6rem;
  display: flex;
  width: 100%;
`;

const Track = styled('div')`
  height: 1.2rem;
  width: 100%;
  box-shadow: 0 0 0 0.3rem ${({theme}) => theme.palette.grey.dark};
  border-radius: 2rem;
  background: ${({values, theme, min, max}: any) => (
  getTrackBackground({
    values,
    colors: [theme.palette.primary.light, theme.palette.light],
    min,
    max
  })
)};
  align-self: center
`;

const Thumb = styled('div')<RailProps>`
  z-index: 1 !important;
  height: 3rem;
  width: 2.5rem;
  border-radius: 0.4rem;
  background-color: ${({isDragged, theme}) => isDragged ? theme.palette.grey.dark : theme.palette.light}; 
  display: flex;
  justify-content: center;
  font-weight: bold;
  align-items: center;
  box-shadow: 0 0 0 0.3rem ${({theme}) => theme.palette.grey.dark};
  color:  ${({isDragged, theme}) => isDragged ? theme.palette.light : theme.palette.grey.dark};
`;

const Tips = styled('span')`
  color: ${({theme}) => theme.palette.grey.light};
  position: absolute;
  bottom: -1rem;
  right: 0;
`;

const RangeWrapper = styled('div')`
  margin: 2rem 1.2rem;
  position: relative;
`;

const MinMax = styled('div')`
  position: absolute;
  width: 100%;
  display: flex;
  justify-content: space-between;
  top: 50%;
  left: 0;
  transform: translateY(-50%);
  padding: 0.5rem;
  font-size: 1rem;
`;

const InputRange: React.FC<RangeProps> = ({field, min, max, step, onChange}) => {
  const [values, setValues] = useState<number[]>([min]);

  const changeValues = (values: number[]) => {
    setValues(values);
    onChange(values[0])
  };

  return (
    <Fragment>
      <FrequencyLabel htmlFor={field.name}>What will your <strong>watering frequency</strong> be?</FrequencyLabel>
      <RangeWrapper>
        <Range
          values={values}
          min={min}
          max={max}
          step={step}
          onChange={changeValues}
          renderTrack={({props, children}) => (
            <Rail onMouseDown={props.onMouseDown} onTouchStart={props.onTouchStart}>
              <Track ref={props.ref} {...props} {...{values, min, max}}>{children}</Track>
            </Rail>
          )}
          renderThumb={({props, isDragged}) => (
            <Thumb{...props} isDragged={isDragged}>
              {props["aria-valuenow"]}
            </Thumb>
          )}
        />
        <MinMax>
          <span>{min}</span>
          <span>{max}</span>
        </MinMax>
        <Tips>in days</Tips>
      </RangeWrapper>
    </Fragment>
  );
};

export default InputRange;
