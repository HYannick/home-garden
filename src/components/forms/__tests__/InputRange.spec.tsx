import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup } from '@testing-library/react';
import { renderWithTheme } from '../../../theme-wrapper';
import InputRange from '../InputRange';

afterEach(cleanup);
describe('InputRange', () => {
  const props = {
    field: {
      name: 'puke',
    },
    type: 'text',
    label: 'Should I puke?',
    min: 2,
    max: 10,
    step: 1,
    onChange: jest.fn()
  };
  it('should render properly', () => {
    const { container } = renderWithTheme(
      <InputRange {...props}/>,
    );
    expect(container).toMatchSnapshot();
  });
});
