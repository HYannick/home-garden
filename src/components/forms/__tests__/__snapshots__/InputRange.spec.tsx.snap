// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRange should render properly 1`] = `
<div>
  <label
    class="css-he3bhv"
    for="puke"
  >
    components.frequency.label
  </label>
  <div
    class="css-18uljbk"
  >
    <div
      class="css-98r97g"
    >
      <div
        class="css-1wypj0"
        max="10"
        min="2"
        style="cursor: pointer;"
        values="2"
      >
        <div
          aria-valuemax="10"
          aria-valuemin="2"
          aria-valuenow="2"
          class="css-1r0bhv1"
          draggable="false"
          role="slider"
          style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(NaNpx, NaNpx);"
          tabindex="0"
        >
          2
        </div>
      </div>
    </div>
    <div
      class="css-yvyril"
    >
      <span>
        2
      </span>
      <span>
        10
      </span>
    </div>
    <span
      class="css-1ia2zn2"
    >
      components.frequency.in_days
    </span>
  </div>
</div>
`;
